### PR TITLE
Add continuous testing on Windows using GitHub Actions

### LINKED ISSUE #8666: Continuous testing on Windows
So we've been getting occasional PRs saying that they make this or that spec pass on Windows. That made me curious whether the status of specs passing/failing is being tracked anywhere.

@straight-shoota said that there hasn't been such a concerted effort, or even a list of specs that are expected to pass under the current status.
Well, not until he implemented this PR :slightly_smiling_face: https://github.com/crystal-lang/crystal/pull/8664

Then, how can we actually run these?
The current instructions https://github.com/crystal-lang/crystal/wiki/Porting-to-Windows require steps on Linux before getting to Windows.

I was inspired by the multi-stage approach of CircleCI, and indeed it allows running 2 sequential jobs on these different systems and pass files among them.

My work in progress: https://github.com/oprypin/crystal/blob/ci/.circleci/config.yml

I'm just following the instructions from the wiki quite exactly. CircleCI even has MSVC and buildtools preinstalled :slightly_smiling_face: 

Sadly, CircleCI provides Windows only 250 minutes / week. And I've already hit that limit for myself...
https://circleci.com/gh/oprypin/crystal

### PR DESCRIPTION
**Pull request by oprypin**
This is a pipeline using a Linux job that cross-compiles for Windows, followed by an actual Windows job linking and running that.
    
Based on https://github.com/crystal-lang/crystal/wiki/Porting-to-Windows/30956f45a4e739204881694ee6c92f4fb0bc5395

Note that all the actual commits from this PR are prefixed with "Win CI:", the rest is merges of pull requests that this relies on.
This PR creates and touches *only* the file ".github/workflows/win.yml".

**[Example in action](https://github.com/oprypin/crystal/actions?query=branch%3Agh-ci+is%3Asuccess)**

Fixes #8666.

### CODE REVIEW

**Comment by straight-shoota** [comment]
This is really amazing! I almost didn't expect to get CI working on windows so quickly.

A few remarks:
* In order to provide the same behaviour as the existing CI workflows, it would be nice if we could build stdlib specs with both latest and master compiler. I.e. before and after the build crystal step.
* The versions for libgc and libatomic_ops in the guide are outdated. I didn't test it yet, but I suppose the latest releases should work as well. That's v8.0.4 and v7.6.10 as used for the linux compiler build.
* I haven't looked deeply into GitHub actions but I figure we could extract building the library into a re-usable action step? I think that would help to separate concerns and keep the CI config more clear. Alternatively, a separate PowerShell script would also work and even provide the benefit that it would be easily applicable for local installs.

**Comment by oprypin** [comment]
1. Sure, that can be done as a follow-up if deemed necessary. To me it seems like it's just extra runtime for no benefit currently. It's being checked on other systems, and the Windows build doesn't satisfy even weaker guarantees than this one.
2. Yes, but note that these versions are *newer* than what's used in Crystal's official Docker image (Older Ubuntu). Still, I tried updating, but it did not "just work": building bdwgc on Windows has been reworked.
3. * Seems like that ability is a huge jump in difficulty, i.e. requires writing JavaScript code.
   * The separate script already exists and is linked from the wiki, but I found that GitHub Actions config was much more concise and expressive. Generally, grouping actions to a script makes the UI of builds less granular, though it's not so important here.

**Comment by RX14** [comment]
> building bdwgc on Windows has been reworked

It's actually been fixed: `ntwin32.mak` is no longer required. I'd prefer to move to the cmake build for bdwgc if possible.

**Comment by oprypin** [comment]
It's possible to merge this and then update that.

**Comment by oprypin** [comment]
Actually I'm convinced by your argument that it's much simpler now.
Willing to try, but for now getting linking errors, maybe you can advise.
https://github.com/oprypin/crystal/runs/385894088#step:8:2

**Comment by RX14** [comment]
@oprypin ah, it's fixed on latest `bdwgc` but not the release: https://github.com/ivmai/bdwgc/blob/7dcf6be3a74335349aab5bd38fc96594761e5780/CMakeLists.txt#L299

**Comment by oprypin** [comment]
No, didn't work either.
https://github.com/oprypin/crystal/runs/387467545
I am not pursuing this further.
Other than the remaining tiny dependency PR, this is ready.

**Comment by oprypin** [comment]
Well, it can be merged even without #8674, it'll just fail 1 spec initially :>
So fully ready, actually.

**Comment by RX14** [comment]
@oprypin oh! forgot `cflags=-DDONT_USE_USER32_DLL` in the `nmake` as well. Sorry, but I'm sure that that will fix it.

**Comment by oprypin** [comment]
Nope :upside_down_face: https://github.com/oprypin/crystal/runs/387628797#step:9:40

**Comment by RX14** [comment]
ffs

File: .github/workflows/win.yml [hunk 1]
```diff
@@ -0,0 +1,94 @@
+name: Windows CI
+
+on: [push, pull_request]
+
+jobs:
+  linux-job:
+    runs-on: ubuntu-latest
+    container: crystallang/crystal:latest
```
Comment (bcardiff) [inline .github/workflows/win.yml]:
```suggestion
    container: crystallang/crystal:0.32.1
```

It's better to keep the exact version used and update them upon release.

**Comment by oprypin** [comment]
OK but so many specs have regressed since 😞


https://github.com/oprypin/crystal/runs/414564931?check_suite_focus=true

**Comment by oprypin** [comment]
Ah no nevermind, it's just the old state of the repository because I'm not rebasing my branch.

Good to go.

But I *have* dropped the commit that allows Actions to also run on pull requests, to be on the safe side.

**Comment by straight-shoota** [comment]
> But I have dropped the commit that allows Actions to also run on pull requests, to be on the safe side for now.

I'd actually prefer running it on pull requests to immediately see whether the changes would break on windows. It's acceptable if builds occasionally break. That's the same with `preview_mt`.

**Comment by bcardiff** [comment]
Let's see how it goes :-) Maybe if the free tier is not enough we might need to limit branch names.

Thanks @oprypin and others :-)

**Comment by oprypin** [comment]
https://github.com/crystal-lang/crystal/runs/415569434 :tada:

**Comment by j8r** [comment]
We could do the same for Alpine aarch64 :pray:

**Comment by bcardiff** [comment]
@j8r You mean #7420?

**Comment by straight-shoota** [comment]
`aarch64`

**Comment by straight-shoota** [comment]
@oprypin Didn't you say the actions wouldn't be triggered on pull requests? Am I missing something here: https://github.com/crystal-lang/crystal/pull/8683/checks?check_run_id=415666174
There's `on: [push, pull_request]` in the config file.

**Comment by oprypin** [comment]
Oh sorry, I re-added that, per request

**Pull request merged**